### DISCUSSION TITLE
How to exclude empty types when using fragments?

### DISCUSSION THREAD

**Discussion by gopeter** [dotansimha/graphql-code-generator#5567] [Q&A] [answered]
I'm building an app with a Craft CMS GraphQL API and Next.js and have defined this query:

```graphql
query GetHomepage($language: String!) {
  homepage: entry(uri: "__home__", site: [$language]) {
    ... on homepage_homepage_Entry {
      id
      title
      pageContent {
        ... on pageContent_wysiwyg_BlockType {
          wysiwyg
        }
      }
    }
  }
}
```

Looking at this query, it's clear that the type can only be `homepage_homepage_Entry`. But the code generator generates this one: 

```ts
export type GetHomepageQuery = {
	__typename ? : 'Query',
	homepage: Maybe <{
		__typename ? : 'pages_pages_Entry'
	} | {
		__typename ? : 'homepage_homepage_Entry',
		id: Maybe < string > ,
		title: Maybe < string > ,
		pageContent: Maybe < Array < Maybe < {
			__typename ? : 'pageContent_wysiwyg_BlockType',
			wysiwyg: Maybe < string >
		} >>>
	} > 
};
```

So I'm not able to access my data just like this: `homepage.pageContent`, because `pageContent` does not exist here: `{ __typename ? : 'pages_pages_Entry' }`.

Is there anything that I can configure to remove `{ __typename ? : 'pages_pages_Entry' }` from the generated type? For sure `entry` could be of type `pages_pages_Entry`, but I do not query any fields from it and therefore do not need this Union here.

**Comment by dotansimha** [comment, marked as answer]
I'm not sure that `homepage_homepage_Entry` is the only type possible. How does `entry` field is defined? 
Basically, codegen tried to reflect the whole possible response from the server, based on the schema. If there is a need for a fragment - it means that you might have interface/union, and your schema might return other types as well.

**Reply by gopeter** [reply]
Hmmmm, okay. Indeed, the _schema_ has more than one possible type for `entry`, but looking at our _query_ itself, it would be great if codegen could see that we query just one type and therefore does not need to expose all possible types for `entry`, even if we don‘t query them.

**Reply by dotansimha** [reply]
But what if in runtime, the server will return you a different type? That makes the type-safety redundant. If your schema is able to return more than just a single type, your client code should adjust to that. 
If you don't query something - the server can still return it. 
In this specific case - it can return either `pages_pages_Entry` or `homepage_homepage_Entry`. It's better to check it in the client than trying to workaround that, since if it will be returned and you'll ignore it- you'll get an empty object.

**Reply by gopeter** [reply]
Hmmm, not sure I understand what you mean, since the client just asks for one specific type – why should the server return a different type? 🤔 Since I can‘t query types which aren‘t defined (because of Graphql‘s type safety). Maybe I‘m missing something, but if I ask for `homepage_homepage_Entry` explicitly, I do not expect that I could get a `pages_pages_Entry`

**Reply by dotansimha** [reply]
The client doesn't ask for a specific type. For the client it's: `if the type homepage_homepage_Entry is returned, give me these fields: ...`. Setting inline fragment doesn't add a filter for the returned data.

**Reply by gopeter** [reply]
But ... 

```
homepage: entry {
  ... on homepage_homepage_Entry {
    id
    title
  }
}
```

... won't return anything if the type is not `homepage_homepage_Entry`.

I still don't really get why the above query could be anything other than:

```
export type GetHomepageQuery = {
    __typename ? : 'homepage_homepage_Entry',
    id: Maybe < string > ,
    title: Maybe < string > ,
};
```

Well, I don't think that this is something that can be solved easily (or rather be solved at all), so thanks for this great tool and thanks for your answers! I think I have to create some TypeScript helpers which extracts the part of the returned typed with the typename I expect. [Something like this](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcBiEJwLxwIJRQCGiAPAFBxVwDel1VAlgCYBccAzjFIwHYDmAbnoMA+qKQpeRALbB2AcgAiRGEQCMCkXAC+2gD61tTNnF4BXGQCNgUYQyrjJwaXMUq1AJi0M9VAHzC5AD0wQjIqADujAA2MXA2tHAs7Fw8AoJwThGu8nDKqhoKuuTOcACyiACK5raIHho4cACiINxEAMYwpBgQANoW1rYAugA0SdlSsnkARA3qM7r+5CFhZdFxCag0yaaDNnZZEjnT7oXeemWVNXUNnk2t7V09mAOWB2MTx1NucHPnMx0yyAA).

**Reply by dotansimha** [reply]
As I said before, inline fragments are not filters. They are just a specification of the fields you wish to get, in case that type is being returned. 


If you have:

```graphql
interface Node {
  id: ID!
}

type A implements Node {
  id: ID!
  a: String
}

type B implements Node {
   id: ID!
   b: String
}

type Query {
  node: Node!
}
```

And you do:

```graphql
query {
   node {
      ... on A {
         id
         b
      }
   }
}
```

The possible return value for that query is `{ __typename: "B" } | { __typename: "A", id: string, b: string }`. The server can still return `B` here, because the fragment is not a filter. 

The logic in the server *might* return only `A`, but the schema declares that it can return any `Node`, so you should handle this case as well. 

Your options here as I can see it:

1. Ignore all types that you don't care about - but note that you might get it anyway since it's not a filter.
2. If your backend and resolvers know to filter based on fragments specified (using `info` parameter) - then you can safely assume that filter will work. 
3. Change your schema to return a concrete type and not an abstract type.

**Comment by zo-ly** [comment]
I am upgrading the codegen version in my project from 1.1.3 to 2 and have encountered a lot of typescript compatibility issues. 

My problem is similar to this discussion. I found that the v1.1.3 can exclude these empty types, but the v2 can't. Is there any way now to exclude these empty types? 

Here is the [code demonstration](https://github.com/zo-ly/graphql-codegen-test). Thank you very much for this tool.
